src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -106,37 +107,118 @@ class BillingTxCountMetricConsumerStrategy(ProcessingStrategy[KafkaPayload]):
+                    extra={"offset": metrics_msg.offset},
+                )
+
+        self._ongoing_billing_outcomes.popleft()
+        mapping = {metrics_msg.partition: Position(metrics_msg.next_offset, datetime.now())}
+        self._commit(mapping, force)
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+
+        while self._ongoing_billing_outcomes:
+            now = time.time()
+            time_left = deadline - now if deadline else None
+
+            if time_left is not None and time_left <= 0:
+                logger.warning(
+                    f"join timed out, items left in the queue: {len(self._ongoing_billing_outcomes)}"
+                )
+                break
+
+            self._process_metrics_billing_bucket(time_left, force=True)
+
+        self._close_producers()
 
     def terminate(self) -> None:
         self.close()
 
+        if len(self._ongoing_billing_outcomes) > 0:
+            logger.warning(f"terminated, items dropped: {len(self._ongoing_billing_outcomes)}")
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            if ongoing_work:
+                ongoing_work.cancel()
+
+        self._close_producers()
+
+    def _close_producers(self) -> None:
+        for producer in self._producers.values():
+            producer.close()
+
     def close(self) -> None:
-        self.__closed = True
+        self._closed = True
 
     def submit(self, message: Message[KafkaPayload]) -> None:
-        assert not self.__closed
-        self.__messages_since_last_commit += 1
+        assert not self._closed
 
-        payload = self._get_payload(message)
-        self._produce_billing_outcomes(payload)
-        self._mark_commit_ready(message)
+        bucket_payload = self._get_bucket_payload(message)
+        quantity = self._count_processed_transactions(bucket_payload)
 
-    def _get_payload(self, message: Message[KafkaPayload]) -> MetricsBucket:
+        if quantity < 1:
+            # We still want to commmit buckets that don't generate billing
+            # outcomes, since the consumer has already processed them. To keep
+            # the offset order when committing, we still need to go through the
+            # futures queue.
+            future = None
+        else:
+            future = track_outcome_custom_publish(
+                self._produce,
+                org_id=bucket_payload["org_id"],
+                project_id=bucket_payload["project_id"],
+                key_id=None,
+                outcome=Outcome.ACCEPTED,
+                reason=None,
+                timestamp=datetime.now(),
+                event_id=None,
+                category=DataCategory.TRANSACTION,
+                quantity=quantity,
+            )
+
+        self._ongoing_billing_outcomes.append(MetricsBucketBilling(future, message))
```

Comment:
I think you are not applying any backpressure on the queue if the producer slows down for whatever reason.
Imagine for a few seconds your producer becomes slower. Now your queue grows unbounded. This can send your consumer into out of memory which is one of those conditions where the consumer is not terminated cleanly. 

Even before going out of memory you are likely going to fill the buffer of the producer downstream (by default 10000 messages if I remember correctly). When you reach 10k unacknowledged callbacks or pending callbacks the producer will fail. 

You should cap your queue in size. When it is full `raise MessageRejectedException`. That indicates to the Arroyo processor that you cannot accept the message and will retry. Obviously you need to do this check before you try to produce (at the beginning of the `submit` method).

Or, better, use the `ProduceAndCommit` strategy that does this for you, so you do not have to care about backpressure. It is documented here https://getsentry.github.io/arroyo/strategies.html#arroyo.processing.strategies.produce.ProduceAndCommit